Implementation of Hercules Ultimate Storage System (HUSS)

jasonch35 opened the pull request:
<!-- Before you continue, please change "base: stable" to "base: master" and
     enable the setting "[√] Allow edits from maintainers." when creating your
     pull request if you have not already enabled it. -->

<!-- Note: Lines with this <!-- syntax are comments and will not be visible in
     your pull request. You can safely ignore or remove them. -->

### Pull Request Prelude

<!-- Thank you for working on improving Hercules! -->
<!-- Please complete these steps and check the following boxes by putting an `x`
     inside the [brackets] before filing your Pull Request. -->

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

<!-- Describe the changes that this pull request makes. -->

- Implementation of the Hercules Ultimate Storage System
- Storages can now be created through a configuration file that describes their attributes.
- Integrated storage_constants (Constant) into the configuration so that it can be assigned dynamically
- Capacity is capped by `MAX_STORAGE` at config reading

**Storage Configuration**
```
{
        Id: (int)			Unique Identifier
        Name: (string)			Name of the storage sent to the client.
        Constant: (string)		Storage constant to be used by scripts
        Capacity: (int)			Maximum capacity of the storage.
}
```
**Script Command**
`openstorage(<storage_constant>{, <storage_mode>});`

**Storage Modes**

- STORAGE_ACCESS_VIEW:		View storage only.
- STORAGE_ACCESS_GET:		Allow retrieving items from storage.
- STORAGE_ACCESS_PUT:		Allow depositing items into storage.
- STORAGE_ACCESS_ALL:		(Default mode) Full access for viewing, retrieving, and depositing items.

**All atcommands utilizing storage has been updated as well:**
- `@storage <storage name/id>`
- `@storeall <storage name/id>`
- `@clearstorage <storage name/id>`
- `@storagelist <storage name/id>`

**Issues addressed:** <!-- Write here the issue number, if any. -->

- [x] #1763
- [x] #2724
- [x] #3302

Upon testing the following has been fixed as well:
- [x] #822
- [x] #1659
Thank you @kyeme for pointing it out.

**Credits to**: @sagunkho and @dastgirp

Let's gooooo let's push this through!
 
<!-- You can safely ignore the links below:  -->


[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

jasonch35 committed:
Implementation of the Hercules Ultimate Storage System

This commit includes:
- HUSS main implementation
- openstorage script command update
- @storage atcommand update
- @clearstorage atcommand update
- @storeall atcommand update
- @storagelist atcommand update

jasonch35 committed:
Add SQL files for main db and upgrades

jasonch35 committed:
Add data on message tables for the updated atcommands

jasonch35 committed:
Storage configuration

- There can be an unlimited amount of storages and limits.
- All setting names are case-sensitive and must be keyed accurately.
- The storage capacity will default to MAX_STORAGE in mmo.h if it is set higher
- Storage ID 1 is the default (official) storage for accounts.
- Integrated storage_constants (Constant) into the configuration so that it can be assigned dynamically

jasonch35 committed:
Update NPC scripts to comply with new command

jasonch35 committed:
Update script and atcommand documentation

jasonch35 committed:
Update HPM Hooks

jasonch35 committed:
Show warning if storage capacity is over MAX_STORAGE

jasonch35 commented:
> Thank you @jasonch35 hope this one will get implemented.
> Also i want to ask about Storage Limit its still 600 correct?

Yes, Limit is capped by MAX_STORAGE which is 600 by default.

guilherme-gm reviewed (changes requested):
Thank you for putting the effort in fixing the previous PRs!
I did not try running it yet, but from the code review, I got those questions/points that may need to be addressed.

jasonch35 commented on src/map/storage.c:
```diff
@@ -414,11 +515,12 @@ static int storage_storagegettocart(struct map_session_data *sd, int index, int
 	nullpo_retv(sd);
+	Assert_retv(sd->storage.current > 0);
 
-	Assert_retv(sd->storage.received == true);
```
This assertion is removed so `storage->ensure` always runs depending on player's current storage id, ensuring correct storage id/data being loaded for the character regardless being flagged true or false.
 
And replaced with `return` statement if `storage->ensure` returns with NULL.

guilherme-gm commented on src/map/storage.c:
```diff
@@ -277,13 +357,22 @@ static int storage_delitem(struct map_session_data *sd, int n, int amount)
 
-	if (sd->storage.aggregate > MAX_STORAGE)
+	if ((sd->storage.access & STORAGE_ACCESS_PUT) == 0) {
+		clif->delitem(sd, index, amount, DELITEM_NORMAL);
+		clif->additem(sd, index, amount, 0);
+		return 0;
```
oh, cool! so this is also the start to fix another issue and the actual official behavior.

thanks!

guilherme-gm commented on src/map/storage.c:
```diff
@@ -345,26 +436,35 @@ static int storage_add_to_inventory(struct map_session_data *sd, int index, int
+	nullpo_ret(stst = storage->get_settings(stor->uid));
 
-	if (sd->storage.aggregate > MAX_STORAGE)
+	if ((sd->storage.access & STORAGE_ACCESS_PUT) == 0) {
+		clif->delitem(sd, index, amount, DELITEM_NORMAL);
+		clif->additem(sd, index, amount, 0);
+		return 0;
+	}
```
I did not test, but another question here: `clif->delitem` and `clif->additem` are related to player inventory, in this case, shouldn't we use the calls for Cart instead (`clif->cart_additem` / `clif->cart_delitem`)? since this is a move from cart to storage

jasonch35 replied:
Just tested. You're right, should be `clif->cart_delitem` / `clif->cart_additem`. Changed.

jasonch35 committed:
Use cart function when storing from cart

jasonch35 committed:
Indexed storage_id column from storage sql table

jasonch35 committed:
Char-server to ignore storage sizes and hardcoded to MAX_STORAGE instead

Reverted to local variables since its not needed to allocate anymore.

jasonch35 committed:
Moved storage_settings struct from mmo to storage header

jasonch35 committed:
Added access modes argument for storage->open

Adjusted callers

jasonch35 committed:
Refactored atcommands logic

-Direct initialization for struct instead to NULL
-Removed `intval` variable: copy value directly to `storage_id`
-Removed unnecessary message table comments
-Moved stst NULL check outside of if condition

jasonch35 committed:
Missed name parameter in clif.c

jasonch35 committed:
Fixed invalid storage(0) assertion

jasonch35 committed:
Initialization logic fix

+ few cleanups

jasonch35 committed:
Indentation/space fix

jasonch35 committed:
PCCHECK_STORAGE check fix

jasonch35 committed:
Proper code styling for asterisks

jasonch35 committed:
Removed unnecessary cleaning of VECTOR

jasonch35 committed:
Removed full storage checks for functions that calls storage->additem()

jasonch35 committed:
Set constant at the end of every iteration at config reading

Also added handler for when storage capacity is invalid or set below 1

jasonch35 committed:
Update HPM Hooks

jasonch35 committed:
Messagetable update

-Added @storagelist error message and rearranged
-Matched MSGTBL_STORAGE_NOT_LOADED to its messages.conf text

jasonch35 commented on sql-files/main.sql:
```diff
@@ -975,6 +976,7 @@ INSERT IGNORE INTO `sql_updates` (`timestamp`) VALUES (1653155461); -- 2022-05-2
 CREATE TABLE IF NOT EXISTS `storage` (
   `id` INT UNSIGNED NOT NULL AUTO_INCREMENT,
   `account_id` INT UNSIGNED NOT NULL DEFAULT '0',
+  `storage_id` INT UNSIGNED NOT NULL DEFAULT '1',
```
c5e746ae7976ce407dde571ec5f2c2ad2fc1915f

jasonch35 commented on src/common/msgtable.h:
```diff
@@ -1429,7 +1437,7 @@ enum msgtable_messages {
 
 	// @storage
-	/** Storage opened. */
+	/** %s opened. */
```
03a7aa0f154467717346b7359fe5024d8185ca80

jasonch35 commented on src/char/int_storage.c:
```diff
@@ -41,24 +41,26 @@ static struct inter_storage_interface inter_storage_s;
+	int i = 0, j = 0;
 	if (VECTOR_LENGTH(cp.item) > 0) {
+
+		int *delete = aCalloc(cp_size, sizeof(int));
```
Reverted to local variables since we hardcoded to a fixed MAX_STORAGE anyway.

jasonch35 commented on src/char/int_storage.c:
```diff
@@ -41,24 +41,26 @@ static struct inter_storage_interface inter_storage_s;
+	
+	int cp_size = inter_storage->fromsql(account_id, storage_id, &cp, 0);
+
+	bool* matched_p = aCalloc(VECTOR_LENGTH(p->item), sizeof(bool));
```
Same as above

jasonch35 commented on src/char/int_storage.c:
```diff
@@ -194,10 +201,14 @@ static int inter_storage_fromsql(int account_id, struct storage_data *p)
+		for (i = 0; SQL_SUCCESS == SQL->NextRow(inter->sql_handle); ++i) {
 			struct item item = { 0 };
+
+			if (storage_size > 0 && i > storage_size)
```
Same as above.

jasonch35 commented on src/map/atcommand.c:
```diff
@@ -874,6 +874,40 @@ ACMD(speed)
+			return false;
+		}
+		storage_id = intval;
+	} else if (*message != '\0' && sscanf(message, "%23s", storage_name) == 1) {
```
8baf55b14be0d5a1e64e37d7c3c9b248e95da9b1

jasonch35 commented on src/map/atcommand.c:
```diff
@@ -5635,24 +5671,55 @@ ACMD(dropall)
+			return false;
+		}
+		storage_id = intval;
+	} else if (*message != '\0' && sscanf(message, "%23s", storage_name) == 1) {
```
8baf55b14be0d5a1e64e37d7c3c9b248e95da9b1

jasonch35 commented on src/map/atcommand.c:
```diff
@@ -5635,24 +5671,55 @@ ACMD(dropall)
 ACMD(storeall)
 {
+	char storage_name[NAME_LENGTH] = "";
+	int storage_id = 0, intval = 0;
```
Remove `intval` altogether. And copy value directly to storage_id, is this okay?

jasonch35 commented on src/map/chrif.c:
```diff
@@ -264,8 +264,12 @@ static bool chrif_save(struct map_session_data *sd, int flag)
-		sd->state.storage_flag = STORAGE_FLAG_CLOSED; //Force close it.
+	if (flag != 0 && sd->state.storage_flag != STORAGE_FLAG_CLOSED) {
+		if (sd->state.storage_flag == STORAGE_FLAG_NORMAL)
+			storage->close(sd);
```
storage.c calls `chrif->save(sd, 0)` which is `flag == 0`. I don't think it will. I might be wrong, though. What's the best way to do this?

MishimaHaruna replied:
You're right, flag is 0 so there's no infinite recursion. I'll still call chrif->save twice, but I can't think of anything better right now

jasonch35 commented on src/map/chrif.c:
```diff
@@ -264,8 +264,12 @@ static bool chrif_save(struct map_session_data *sd, int flag)
+		if (sd->state.storage_flag == STORAGE_FLAG_NORMAL)
+			storage->close(sd);
+		else if (sd->state.storage_flag == STORAGE_FLAG_GUILD)
+			gstorage->close(sd);
```
Same as above

jasonch35 commented on src/map/clif.c:
```diff
@@ -3290,7 +3296,7 @@ static void clif_storageItems(struct map_session_data *sd, enum inventory_type t
 #endif
 #if PACKETVER >= 20120925 && PACKETVER_RE_NUM < 20180829 && PACKETVER_ZERO_NUM < 20180919 && PACKETVER_MAIN_NUM < 20181002
-		safestrncpy(storelist_normal.name, "Storage", NAME_LENGTH);
+		safestrncpy(storelist_equip.name, name, NAME_LENGTH);
```
973f6f6b9498e9bd4000d51c938addd9d2432f39

jasonch35 committed:
Clif update storage amount window to capacity

jasonch35 commented on src/map/clif.c:
```diff
@@ -9694,14 +9700,24 @@ static void clif_serviceMessageColor(struct map_session_data *sd, uint32 color,
+
+		clif->storageList(sd, VECTOR_DATA(stor->item), VECTOR_LENGTH(stor->item));
+
+		clif->updatestorageamount(sd, stor->aggregate, MAX_STORAGE);
```
You're right. Client updates to wrong max capacity with this run.

jasonch35 commented on src/map/pc.c:
```diff
@@ -10599,21 +10600,25 @@ static int pc_checkitem(struct map_session_data *sd)
-		if (sd->itemcheck & PCCHECKITEM_STORAGE && sd->storage.received == true) {
-			for (i = 0; i < VECTOR_LENGTH(sd->storage.item); i++) {
-				struct item *it = &VECTOR_INDEX(sd->storage.item, i);
+		if ((sd->itemcheck & PCCHECKITEM_STORAGE) != 0) {
```
e38f415732ccc89c1faa4a5575b85e2965738c02

jasonch35 commented on src/map/script.c:
```diff
@@ -12012,9 +12033,13 @@ static BUILDIN(openstorage)
 	}
 
-	storage->open(sd);
+	sd->storage.access = storage_access; // Set storage access level. [Smokexyz/Hercules]
```
Added argument to `storage->open()`

jasonch35 commented on src/map/storage.c:
```diff
@@ -277,13 +358,22 @@ static int storage_delitem(struct map_session_data *sd, int n, int amount)
+	}
 
-	if (sd->storage.aggregate > MAX_STORAGE)
+	if (stor->aggregate >= stst->capacity)
```
I agree. This is not needed.

jasonch35 commented on src/common/mmo.h:
```diff
@@ -644,7 +644,15 @@ struct status_change_data {
+	int uid;                       ///< Storage Identifier.
+	char name[NAME_LENGTH];        ///< Storage Name
+	int capacity;                  ///< Item Capacity.
+};
```
6e485b3c50a288881297f4cd171d35c92ec4df08

Also regenerated HPM for this.

jasonch35 commented on src/map/storage.c:
```diff
@@ -887,6 +993,145 @@ static int storage_guild_storage_quit(struct map_session_data *sd, int flag)
+			continue;
+		}
+
+		if (s_conf.capacity > MAX_STORAGE) {
```
5e3a480fdd7ab11617a45fc885440890ba0827b8

Skipping storage if `< 1`

jasonch35 commented on src/map/storage.c:
```diff
@@ -887,6 +993,145 @@ static int storage_guild_storage_quit(struct map_session_data *sd, int flag)
+	nullpo_retr(false, filename);
+
+	if (!imported)
+		VECTOR_INIT(storage->configuration);
```
I've been thinking. I wanna drop this feature altogether. Since `storage.conf` will be as is almost forever officially similar to other configs. There is no point having additional fields and imports.

jasonch35 replied:
9f08c77a4a4a41c9ea0bd85a9ff678043961edb0

I transferred config read function below the initialization.

jasonch35 commented on src/map/storage.c:
```diff
@@ -887,6 +993,145 @@ static int storage_guild_storage_quit(struct map_session_data *sd, int flag)
+		storage->config_read_additional_fields(t, &s_conf, filename);
+
+		if (imported) {
+			ARR_FIND(0, VECTOR_LENGTH(storage->configuration), i, VECTOR_INDEX(storage->configuration, i).uid == s_conf.uid);
```
Same as above:
I've been thinking. I wanna drop this feature altogether. Since storage.conf will be as is almost forever officially similar to other configs. There is no point having additional fields and imports.

jasonch35 replied:
9f08c77a4a4a41c9ea0bd85a9ff678043961edb0

Added check on the first duplicate check.

jasonch35 committed:
Corrected documentation

jasonch35 committed:
Storage config reading

- Check duplicate for non imported storage ID and overwrite when imported ID exist
- Move vector initialization on do_init_storage and transferred config reading after it

vBrenth commented:
I know its Holidays but please give us this, we've been waiting for years. (officially)
@MishimaHaruna

guilherme-gm commented on src/map/storage.c:
```diff
@@ -96,37 +99,115 @@ static void do_reconnect_storage(void)
+	const struct storage_settings *stst = storage->get_settings(stor->uid);
+	nullpo_retr(1, stst);
 
-	if (sd->storage.received == false) {
-		clif->message(sd->fd, msg_sd(sd, MSGTBL_STORAGE_NOT_LOADED)); // Storage has not been loaded yet.
-		return 1;
-	}
```
is removing the received check intended? from Haru message, he asked whether turning the message into assertion was intended (as an assertion doesn't inform the user why this happened).

But now, we would allow a not received storage to be "open".

jasonch35 replied:
Yes, I rewritten it as an assertion error but as Haru pointed out, it is no longer needed as the script and atcommands already errors out when called if storage is not loaded yet in intif_parse_account_storage

guilherme-gm replied:
I think that if script/atcommand is already checking, it makes sense to switch from a message to an assert, but I would still keep the assert, because this is the real "entry point" to storage functions and we are expecting it to always have a received storage. This guards us from future issue if someone calls storageopen in a new place and forgets to add the received check.

For example, let's say official adds a new button to the UI that opens the storage wherever you are, I could end up calling storageopen from this new clif function without checking for received because I did not know/forgot.

It would be ok to give a "weird experience" like not working and not giving info to user, and generating an error in console (due to assert) that we would fix later on with proper checks. But it would be a big issue if the unreceived storage simply opened with bad data and later on overwrote the real storage, or the items added at this time disappeared.

If I understood, from Haru message, he was just checking whether the change from user message to "console error" (with assert) was intended, not that it was no longer needed, but I may be wrong.

@MishimaHaruna may you share your thoughts on this?

MishimaHaruna replied:
It's as @guilherme-gm said. Please keep any assertions since they serve, if nothing else, as self-documentation of what a function expects, since it may be called by plugins or by future commands. I was only inquiring whether the removal of the user-facing message was intended

jasonch35 replied:
Re-added assertion. https://github.com/HerculesWS/Hercules/pull/3330/commits/1bb030c6f76ef362e05e5e586d4869922a337396

guilherme-gm commented on src/map/unit.c:
```diff
@@ -3001,11 +3001,16 @@ static int unit_free(struct block_list *bl, enum clr_type clrtype)
+
+			for (int i = 0; i < VECTOR_LENGTH(sd->storage.list); i++) { // Storages
+				VECTOR_CLEAR(VECTOR_INDEX(sd->storage.list, i).item);
+				VECTOR_INDEX(sd->storage.list, i).received = false;
```
(adding here since the line is gone from the final diff)
the vector still needs to be cleared up, because its internal data is dynamically allocated. you just don't have to set `received` to false, as this data will be cleared up at `VECTOR_CLEAR(sd->storage.list);`

guilherme-gm commented on src/map/intif.c:
```diff
@@ -336,31 +342,35 @@ static void intif_parse_account_storage(int fd)
 	storage_count = (payload_size/sizeof(struct item));
 
-	VECTOR_ENSURE(sd->storage.item, storage_count, 1);
+	VECTOR_ENSURE(stor->item, storage_count, 1);
```
Memory manager reported a leak from this, I didn't check what is causing it (not sure if related to the missing VECTOR_CLEAR in unit.c

```
0001 : storage.c line 298 size 352 address 0x0x55976f3717e4
0002 : intif.c line 363 size 440 address 0x0x55977473f1fc
```

jasonch35 replied:
Added https://github.com/HerculesWS/Hercules/pull/3330#discussion_r1894515580, error now gone for me. Thank you!

guilherme-gm commented on src/map/storage.c:
```diff
@@ -200,69 +281,68 @@ static int storage_additem(struct map_session_data *sd, struct item *item_data,
-		VECTOR_PUSH(sd->storage.item, *item_data);
-		it = &VECTOR_LAST(sd->storage.item);
+	if (i == VECTOR_LENGTH(stor->item)) {
+		VECTOR_ENSURE(stor->item, 1, 1);
```
Memory manager reported a leak from this, I didn't check what is causing it (not sure if related to the missing VECTOR_CLEAR in unit.c

```
0001 : storage.c line 298 size 352 address 0x0x55976f3717e4
0002 : intif.c line 363 size 440 address 0x0x55977473f1fc
```

jasonch35 committed:
Removed redundant assignment

jasonch35 committed:
Clear item list for every storage at unit freeing

jasonch35 commented on src/char/int_storage.c:
```diff
@@ -159,44 +158,40 @@ static int inter_storage_tosql(int account_id, const struct storage_data *p)
-
-		VECTOR_ENSURE(p->item, num_rows > MAX_STORAGE ? MAX_STORAGE : num_rows, 1);
+	if (SQL->NumRows(inter->sql_handle) > 0) {
+		VECTOR_ENSURE(p->item, MAX_STORAGE, 1);
```
as Haru stated above https://github.com/HerculesWS/Hercules/pull/3330#discussion_r1864431396, number of items showing in the storage should retrieve actual (_or more_) amount even if its higher than storage capacity and should be hard limited only by MAX_STORAGE. Instead, we let map server to handle the limits. Which I agree in case owners reduce the capacity of a specific storage.

guilherme-gm replied:
I agree with this approach of hard limiting to `MAX_STORAGE` only, what I mean is that the current code would always allocate `MAX_STORAGE` items, even if the storage only has 1 item (not as a limit, but as current slots in use)

Syntax of VECTOR_ENSURE:
`VECTOR_ENSURE(<Target Vector>, <How many empty spaces must be there>, <How much to increase if there is not enough spaces>)`

If you check current master code, you have this:

```
VECTOR_ENSURE(p->item, num_rows > MAX_STORAGE ? MAX_STORAGE : num_rows, 1);
```

Ensure p->item has `num_rows` empty spaces, not exceeding `MAX_STORAGE`. Thus:

Let's say `MAX_STORAGE = 600`
1. if there are 5 items in storage, p->item will be a vector with 5 spaces **(not 600)**
2. if there are 600 items in storage, p->item will be a vector with 600 spaces **(ok)**
3. if there are 800 items in storage, p->item will be a vector with 600 spaces **(hard limited)**


In the new code:

```
VECTOR_ENSURE(p->item, MAX_STORAGE, 1);
```

Ensure p->item has `MAX_STORAGE` empty spaces. Thus:

Let's say `MAX_STORAGE = 600`
1. if there are 5 items in storage, p->item will be a vector with 600 spaces **(595 unused ones)**
2. if there are 600 items in storage, p->item will be a vector with 600 spaces **(ok)**
3. if there are 800 items in storage, p->item will be a vector with 600 spaces **(hard limited)**

I think we don't really need to allocate those 595 unused spaces on case 1. But I agree that we should get everything and hard limit at `MAX_STORAGE`.

jasonch35 replied:
https://github.com/HerculesWS/Hercules/pull/3330/commits/f2c4ffeb60f6049980a1e829b0e44ce8fe574881

I see now, yeah, it shouldn't allocate more than needed.

I want to add though, if server owners decide to decrease `MAX_STORAGE` from 600 to, say, 500. There will be missing items for players that previously have 500+ item count, but only visually. I think this should be fine, though. Any thoughts?

MishimaHaruna replied:
While from an UX perspective it would be better that all the items (even the ones above the limit) are shown, but they can only be retrieved and adding is blocked until usage goes below the limit, I'm ok with hard limiting to `MAX_STORAGE` (and thus not displaying the extra items as you said) if that simplifies the code or the performance by hard limiting the size of some buffers.

Ideally, when over storage capacity, even if it's visually truncating the available items, it should still prevent the player from adding items though. For example, with `MAX_STORAGE = 600`, if player has 610 items, only the first 600 will be displayed. If one is removed (visually only 599 are present, but the total amount is actually 609), it would be best if adding any more was prevented. I understand this is an edge case, so if you feel it would complicate the logic too much, I'll concede.

jasonch35 commented on src/map/script.c:
```diff
@@ -12035,7 +12035,7 @@ static BUILDIN(openstorage)
 
 	sd->storage.access = storage_access; // Set storage access level. [Smokexyz/Hercules]
```
Removed. thank you. please rereview.

jasonch35 committed:
Allocate only required number of rows retrieved from SQL

- Limit rows to MAX_STORAGE, as additional rows are unnecessary.
- Loops based on capped number of rows so we avoid iterating empty items

MishimaHaruna commented:
I think we're almost good to go with this one, the only blocker on my side is the state validation in the storage functions (the missing/removed assertions).

Looking forward to merge it early in the February milestone if possible, so that it can sit on the master branch for a couple of weeks, giving anyone interested a little bit of time to test it in different scenarios before the Feb release at the end of the month.

jasonch35 committed:
Reverted storage availability assertion

jasonch35 commented:
Tested and pushed new reviews. Please rereview @MishimaHaruna @guilherme-gm. Thank you!

guilherme-gm reviewed (approved):
looks good to me. Also performed some in-game tests and the leaks from before are now gone.

kyeme commented:
I tried it ingame and it seems ok, I don't see any problems.
